File: algorithms/2_medium/C/00216_combination_sum_iii/backtrack.py
class Solution:
    def backtrack(self, res, curr_vec, k, n, curr_sum, curr_num):

        if len(curr_vec) == k:
            if curr_sum == n:
                res.append(curr_vec[:])

            return

        for i in range(curr_num, 10):
            if i + curr_sum <= n:
                curr_vec.append(i)
                self.backtrack(res, curr_vec, k, n, curr_sum + i, i + 1)
                curr_vec.pop()


    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        temp = []

        self.backtrack(res, temp, k, n, 0, 1)

        return res

File: algorithms/2_medium/C/00216_combination_sum_iii/backtrack.rs
impl Solution {
    fn backtrack(res: &mut Vec<Vec<i32>>, curr_vec: &mut  Vec<i32>, k: usize, n: i32, curr_sum: i32, curr_num: i32) {

        if curr_vec.len() == k {
            if curr_sum == n {
                res.push(curr_vec.clone());
            }

            return;
        }

        for i in curr_num..10 {
            if i + curr_sum <= n {
                curr_vec.push(i);
                Self::backtrack(res, curr_vec, k, n, curr_sum + i, i + 1);
                curr_vec.pop();
            }
        }
    }

    pub fn combination_sum3(k: i32, n: i32) -> Vec<Vec<i32>> {

        let mut res: Vec<Vec<i32>> = Vec::new();
        let mut temp: Vec<i32> = Vec::with_capacity(k as usize);

        Self::backtrack(&mut res, &mut temp, k as usize, n, 0, 1);

        res
    }
}
